db/impl/consul/lock.go

```diff
@@ -0,0 +1,240 @@
+package consul
+
+import (
+	"reflect"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/contiv/errored"
+	"github.com/contiv/volplugin/db"
+	"github.com/contiv/volplugin/db/jsonio"
+	"github.com/contiv/volplugin/errors"
+	"github.com/hashicorp/consul/api"
+)
+
+// Acquire and permanently hold a lock.
+func (c *Client) Acquire(obj db.Lock) error {
+	// these will be cleared (or refreshed) long before they expire so a 15m ttl is a good way to enforce locks without being too aggressive.
+	_, err := c.doAcquire(obj, 15*time.Minute, true)
+	return err
+}
+
+// AcquireWithTTL holds a lock with an expiration TTL. Attempts only once.
+func (c *Client) AcquireWithTTL(obj db.Lock, ttl time.Duration) error {
+	_, err := c.doAcquire(obj, ttl, false)
+
+	return err
+}
+
+// Free a lock. Passing true as the second parameter will force the removal.
+func (c *Client) Free(obj db.Lock, force bool) error {
+	path, err := obj.Path()
+	if err != nil {
+		return errors.LockFailed.Combine(err)
+	}
+
+	logrus.Debugf("Attempting to free %q by %v", path, obj)
+
+	mylock, ok := c.getLock(path)
+	if !ok {
+		return errors.LockFailed.Combine(errored.New("Could not locate lock"))
+	}
+
+	if !reflect.DeepEqual(obj, mylock.obj) {
+		if force {
+			goto free
+		}
+
+		return errors.LockFailed.Combine(errored.New("invalid lock requested to be freed (wrong host?)"))
+	}
+
+free:
+	select {
+	case <-mylock.monitorChan:
+	default:
+		mylock.lock.Unlock()
```

Comment:
I don't think the re-pointing of mylock is possible with the current architecture; it's never reassigned after it meets the map, so this is quasi-FP with regard to value usage.

So there are two things happening here:

* the monitorchan closes when the lock expires. we don't need to do anything here.
* the alternative is to force the lock free.

Now, we should be able to rely on the consul client's behavior here -- to close the monitorchan if the lock is freed -- which will hit the first case in select and not attempt to double-free. The second access is protected by mutex (everywhere, I hope) by way of the value being gathered during the map access. This also ensures the a rewrite of that lock *to* the map does not affect the current value of this lock.